lib/engine-universal/src/engine.rs

```diff
@@ -79,6 +87,352 @@ impl UniversalEngine {
+        let dynamic_function_trampolines = compilation.get_dynamic_function_trampolines();
+        let data_initializers = translation
+            .data_initializers
+            .iter()
+            .map(wasmer_types::OwnedDataInitializer::new)
+            .collect();
+
+        let frame_infos = compilation.get_frame_info();
+        Ok(crate::UniversalExecutable {
+            function_bodies: compilation.get_function_bodies(),
+            function_relocations: compilation.get_relocations(),
+            function_jt_offsets: compilation.get_jt_offsets(),
+            function_frame_info: frame_infos,
+            function_call_trampolines,
+            dynamic_function_trampolines,
+            custom_sections: compilation.get_custom_sections(),
+            custom_section_relocations: compilation.get_custom_section_relocations(),
+            debug: compilation.get_debug(),
+            trampolines: compilation.get_trampolines(),
+            compile_info,
+            data_initializers,
+            cpu_features: self.target().cpu_features().as_u64(),
+        })
+    }
+
+    /// Load a [`UniversalExecutable`](crate::UniversalExecutable) with this engine.
+    pub fn load_universal_executable(
+        &self,
+        executable: &UniversalExecutable,
+    ) -> Result<UniversalArtifact, CompileError> {
+        let info = &executable.compile_info;
+        let module = &info.module;
+        let local_memories = (module.import_counts.memories as usize..module.memories.len())
+            .map(|idx| {
+                let idx = MemoryIndex::new(idx);
+                (module.memories[idx], info.memory_styles[idx].clone())
+            })
+            .collect();
+        let local_tables = (module.import_counts.tables as usize..module.tables.len())
+            .map(|idx| {
+                let idx = TableIndex::new(idx);
+                (module.tables[idx], info.table_styles[idx].clone())
+            })
+            .collect();
+        let local_globals: Vec<(GlobalType, GlobalInit)> = module
+            .globals
+            .iter()
+            .skip(module.import_counts.globals as usize)
+            .enumerate()
+            .map(|(idx, (_, t))| {
+                let init = module.global_initializers[LocalGlobalIndex::new(idx)];
+                (*t, init)
+            })
+            .collect();
+        let mut inner_engine = self.inner_mut();
+
+        let local_functions = executable.function_bodies.iter().map(|(_, b)| b.into());
+        let function_call_trampolines = &executable.function_call_trampolines;
+        let dynamic_function_trampolines = &executable.dynamic_function_trampolines;
+        let signatures = module
+            .signatures
+            .iter()
+            .map(|(_, sig)| inner_engine.signatures.register(sig.into()))
+            .collect::<PrimaryMap<SignatureIndex, _>>()
+            .into_boxed_slice();
+        let (functions, _, dynamic_trampolines, custom_sections) = inner_engine.allocate(
```

Comment:
I think it is important that we decide where exactly we hold our data, in the store/engine or in the artifact. The benefit for holding everything in the `Artifact` would mean straightforward lifetime management of all the associated structures and data. Holding everything in the store/engine could give us a higher performance headroom but managing the lifetime of the data therein would be way more complicated without effectively reinventing an allocator of some sort.

Holding some data here and some there, however, is the worst of both worlds.